### DISCUSSION TITLE
Variable Contribution/decomposition

### DISCUSSION THREAD

**Discussion by xijianlim** [google/lightweight_mmm#25] [General]
Hi, is there a way to get variable contributions in particular: trend, extra features, base/intercept and media features?

**Comment by pabloduque0** [comment]
Hi there @xijianlim ! thanks for opening up this discussion :)

For media contributions you can rely con [get_posterior_metrics](https://lightweight-mmm.readthedocs.io/en/latest/api.html#lightweight_mmm.lightweight_mmm.LightweightMMM.get_posterior_metrics). Which can also be plotted with [plot_bars_media_metrics](https://lightweight-mmm.readthedocs.io/en/latest/api.html#lightweight_mmm.plot.plot_bars_media_metrics).

We are releasing another plot (hopefully next week) that does the contribution over time of each media variable and the baseline (anything non media: trend, extra features, base/intercept, ...).

However if you are interested in doing custom calculations yourself, everything can be found in the model object in the "trace" attribute. Being `mu` your "prediction" you can calculate the contribution of any variable by multiplying its beta against the data. Feel free to ask follow up questions if you have specific questions over certain calculations.

Also I would be interested in exactly what kind of contributions you are looking to obtain: over time, distributions, point estimates, ... so we can add more functionality in case is missing.


FYI @yukaabe

**Reply by pabloduque0** [reply]
The one I mentioned above is now available: [create_media_baseline_contribution_df](https://lightweight-mmm.readthedocs.io/en/latest/api.html#lightweight_mmm.plot.create_media_baseline_contribution_df)

**Reply by xijianlim** [reply]
That's awesome Pablo! thank you so much. I've got another question related to the response curves but I'll ask it in a separate question.

**Reply by xijianlim** [reply]
@pabloduque0 This is just for reference, but I've used mmm.trace['beta_extra_features'] * extra_features_train to calculate the contribution for each extra feature column.

I hope this is right!

**Reply by pabloduque0** [reply]
Hello @xijianlim ! yes, that will be the correct way to calculate it. Then you can aggregate at the level you desire for the insights you are looking for :)

Also keep in mind scaling if you have done any on the extra features

**Reply by selectabo** [reply]
Hi @pabloduque0 @xijianlim - I'm using the above method to get the coefficients for extra_features. 

e.g. mmm.trace['coef_extra_features'].mean(axis=0) * extra_features_train

However, as you said you need to keep in mind scaling but I think I'm getting it wrong (multiplying coefficients vs the unscaled data) as i'm getting a bigger output than what the 'create_media_baseline_contribution_df' function is producing for baseline.

Is there something different I should be doing to account for scaling?

thanks

**Reply by pabloduque0** [reply]
Hello @selectabo !
What is exactly bigger than the baseline? Do you have multiple extra features? If so, are some of the coefficients positive and others negative?

Could you add a bit more info regarding what you are getting and what would you expect? 
Thanks!

**Reply by selectabo** [reply]
Hi @pabloduque0,

I've got 6 extra features in my model. Some are positive and some are negative.

When I run the create_media_baseline_contibution_df function I get a breakdown of my 4 media channels and a baseline contribution - as an example for one week total sales were 193 with a baseline contribution of 144. I want to get the contribution of each of my extra features - these include things like organic search.

I'm expecting the sum of outputs from the extra features coefficients to be within the range 144 with a remainder being 'baseline'. However, if I use the coefficients for my extra features * the unscaled extra feature value i'm getting over 5000 so I must be doing something wrong.

I'm new to this so i'm pretty sure i'm wrong somewhere!

Thanks

**Reply by pabloduque0** [reply]
My first guess will be that you did not pass the target scaler to `create_media_baseline_contribution_df` and therefore one is scaled and the other one isnt. Could that be the case?

If not let me know and we can try to see what is going on.

**Reply by selectabo** [reply]
I am using the target scaler in create_media_baseline_contribution_df.

"contribution = plot.create_media_baseline_contribution_df(media_mix_model=mmm,target_scaler=target_scaler)"

Should I be using extra features coefficients * scaled extra features inputs? If I do that then they're really small, like less than 1.

**Reply by pabloduque0** [reply]
Although I think you are doing it right, just making sure you are not summing across the samples dimensions, for example doing the mean across samples first:

`mmm.trace['coef_extra_features'].mean(axis=0) * extra_features_train`


Could you show the exact 2-3 lines that you are using that produce the exact output you are referring to? Thanks!

**Reply by selectabo** [reply]
I'm just using those two lines above, to get the contribution for everything and then the features one separately.

contribution = plot.create_media_baseline_contribution_df(media_mix_model=mmm,target_scaler=target_scaler)
features = mmm.trace['coef_extra_features'].mean(axis=0) * extra_features_train

**Reply by pabloduque0** [reply]
@selectabo sorry for the slow response!

The only case where that behaviour can happen is if the seasonality, trend and intercept sum are negative. Then extra features contribution will be higher than the "total baseline".
